bpfd-operator/internal/bpfd-client-helpers.go

```diff
@@ -110,10 +119,167 @@ func LoadTLSCredentials(tlsFiles Tls) (credentials.TransportCredentials, error)
+
+	var decoded []byte
+
+	// StdEncoding can only decode padded string
+	// RawStdEncoding can only decode unpadded string
+	if strings.HasSuffix(strings.TrimSpace(field), "=") {
+		// decode padded data
+		decoded, err = base64.StdEncoding.DecodeString(field)
+	} else {
+		// decode unpadded data
+		decoded, err = base64.RawStdEncoding.DecodeString(field)
+	}
+
+	if err != nil {
+		return
+	}
+
+	parts := strings.SplitN(string(decoded), ":", 2)
+	if len(parts) != 2 {
+		err = fmt.Errorf("unable to parse auth field, must be formatted as base64(username:password)")
+		return
+	}
+
+	username = parts[0]
+	password = parts[1]
+
+	return
+}
+
+// Mimicking exactly what Kubernetes does to pull out auths from secrets:
+// https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/secrets/secrets.go#L29
+func ParseAuth(c client.Client, secretName string) (*ContainerConfig, error) {
+
+	var creds ContainerConfig
+
+	// Lookup the k8s Secret for repository authentication
+	ctx := context.Background()
+	imageSecret := &v1.Secret{}
+	if err := c.Get(ctx, types.NamespacedName{Namespace: v1.NamespaceAll, Name: secretName}, imageSecret); err != nil {
+		return nil, fmt.Errorf("failed image auth secret %s: %v",
+			secretName, err)
+	}
+
+	if containerConfigJSONBytes, containerConfigJSONExists := imageSecret.Data[v1.DockerConfigJsonKey]; (imageSecret.Type == v1.SecretTypeDockerConfigJson) && containerConfigJSONExists && (len(containerConfigJSONBytes) > 0) {
+		containerConfigJSON := ContainerConfigJSON{}
+		if err := json.Unmarshal(containerConfigJSONBytes, &containerConfigJSON); err != nil {
+			return nil, err
+		}
+
+		creds = containerConfigJSON.Auths
+	} else if dockercfgBytes, dockercfgExists := imageSecret.Data[v1.DockerConfigKey]; (imageSecret.Type == v1.SecretTypeDockercfg) && dockercfgExists && (len(dockercfgBytes) > 0) {
+		dockercfg := ContainerConfig{}
+		if err := json.Unmarshal(dockercfgBytes, &dockercfg); err != nil {
+			return nil, err
+		}
+
+		creds = dockercfg
+	}
+
+	return &creds, nil
+}
+
+func BuildBpfdLoadRequest(bpf_program_config_spec *bpfdiov1alpha1.BpfProgramConfigSpec, name string, c client.Client) (*gobpfd.LoadRequest, error) {
 	loadRequest := gobpfd.LoadRequest{
 		SectionName: bpf_program_config_spec.Name,
-		Location:    bpf_program_config_spec.ByteCode,
+	}
+
+	if bpf_program_config_spec.ByteCode.Image != nil {
+		bytecodeImage := bpf_program_config_spec.ByteCode.Image
+		var pullPolicy gobpfd.ImagePullPolicy
+
+		ref, err := reference.ParseNamed(bytecodeImage.Url)
+		if err != nil {
+			panic(err)
+		}
+
+		var username, password string
+		if bytecodeImage.ImagePullSecret != "" {
+			creds, err := ParseAuth(c, bytecodeImage.ImagePullSecret)
+			if err != nil {
+				return nil, err
+			}
+
+			domain := reference.Domain(ref)
+
+			if domain == "docker.io" || domain == "" {
+				domain = "https://index.docker.io/v1/"
+			}
```

Comment:
For future code readers, potentially as years go on, this defaulting behavior could use some explaining comments.